### ISSUE TITLE
DisruptionBudget object to define the max disruption that can be caused to pods

### ISSUE THREAD

**Issue by nikhiljindal** [kubernetes/kubernetes#12611]
Forked from https://github.com/kubernetes/kubernetes/pull/12236.
@davidopp suggested introducing a DisruptionBudget object in https://github.com/kubernetes/kubernetes/pull/12236#discussion_r36825964:

> will it be possible to construct enforceable SLOs if each controller that can cause disruptions has 
> a separate disruption budget spec? (An example of another such controller is rescheduler, but I'm sure there will be others, e.g. maybe one that takes machines offline for kernel upgrades.)
> 
> It seems that we may want a new object that lets you define arbitrary collections of Pods (using label
> selectors) and specifies a correspond disruption budget (maxUnavailable like you have here, max rate of churn, etc.). The Spec for this object would be the budget, and the Status would be the
> number currently down or current churn rate or whatever. This object could be owned by a service that controllers ask permission before they initiate a disruption, or the logic for checking if it is OK to disrupt and updating the status could be implemented in a library that is used by all the controllers that
> cause disruptions.
> 
> How will DisruptionBudget in Deployment work? What if a DisruptionBudget (or whatever) object already exists that covers 
> (some of) these Pods? Does it override the one specified in Deployment? (i.e. you only 
> create DisruptionBudget if it doesn't exist?) Is it an error to specify a disruption budget in Deployment if there is a DisruptionBudget object for (some of) the Pods covered by the Deployment object?

In Deployment, we are currently using a maxUnavailable field. We can eventually switch to using a DisruptionBudget.

cc @bgrant0607

**Comment by davidopp** [comment]
Copying @bgrant0607 's responses to my questions:

> General issue of how to deal with conflicting overlapping sets, similar to multiple controllers. Added to my list of API conventions to document.
>
> Generation of the DisruptionBudget should be optional, though, in case someone did want to create it separately deliberately.
>
> FWIW, internally, self-inflicted disruptions (updates, crash loops, etc.) are not counted against the infrastructure disruption budget.

Last sentence is an excellent point. I think that policy makes a lot of sense, so maybe we wouldn't want Deployment to ever consult the disruption budget. Anyway, not something we need to decide now.

**Comment by davidopp** [comment]
cc/ @thockin (you mentioned something related to this in a meeting today)

**Comment by bgrant0607** [comment]
cc @mml

**Comment by mml** [comment]
Re: not counting self-inflicted wounds against the budget: Yes, but we do still want to count unplanned failures, such as a machine panic or crash.  We need to be able to distinguish these.

"Budget" is a good metaphor for one of the things we usually care about around SLO enforcement: evictions/time.  We can model these with a token bucket.

"Budget" might be a bad metaphor for binary conditions.  One example is shard strength.

Do we have a good handle on the kinds of SLOs users want to see, or cluster admins want to offer?

**Comment by davidopp** [comment]
> Do we have a good handle on the kinds of SLOs users want to see, or cluster admins want to offer?

Not really. Deployment already has maxUnavailable, and I assume we want that. Presumably we want evictions/time as well, as you mentioned. I'm not sure we need anything else, at least for now.

**Comment by bgrant0607** [comment]
Forgiveness is related: #1574

So is service-level readiness, which was discussed in the Component Registration API context #13216.

**Comment by mml** [comment]
Occurred to me while working on #17393 that we should define and enforce an upper bound on min time between evictions, not just so we can reasonably make progress, but also so we don't end up in a degenerate case where the drain controller is trying to establish/maintain 2 years worth of disruption history.

There might be other cases where we need boundaries to make implementation feasible.

**Comment by thockin** [comment]
The issue of control objects with selectors that overlap comes up again and
again, and I am not really satisfied with (my understanding of) the answer.

It came up again on network policies - how do I apply policies to sets of
pods and know that the policies are not in conflict?  It seems easier to
have pods declare which policy they belong to rather than policies declare
which pods they apply to.

Obviously, this doesn't work for RCs and similar.

**Comment by bgrant0607** [comment]
@mml Forgiveness is similar; it also needs a rate.

@thockin 
For controllers the tie-breaking proposal is controllerRef: https://github.com/kubernetes/kubernetes/issues/2210#issuecomment-135588169, #14961

Required label keys could also be used to ensure non-overlap: #15390

However, whether pod specifies policy or vice versa depends on which scenarios we want to facilitate. Related policy discussion: #17097

**Comment by mml** [comment]
We really need to collect user feedback about how much flexibility is needed here.  If a cluster-wide policy is adequate, it can probably be configured via flags, and it will be easier to implement than the fully general idea of a DisruptionBudget object with selectors.

**Comment by davidopp** [comment]
If these are global (or per-namespace) policies as you are advocating, then you presumably need to express them as a fraction of collection size, since a ReplicaSet or Job or whatever with 100 pods can probably tolerate more simultaneous pods down and higher eviction rate than one with 2 pods. Even as a fraction of collection size I'm not sure you can set a good global number, though. At the very least, I suspect some collections will want to opt out of the policy.

Also, we need to track the current state in order to enforce any kind of policy (e.g. how many pods have been disrupted from this collection during the last N minutes). Assuming we want that to be persisted, we need an object anyway, so it seems reasonable to add to the controllers a pointer to a DisruptionBudget object, that contains both the Spec (disruption rate limit and max simultaneously down pods) and the Status (recent disruptions).

**Comment by mqliang** [comment]
@davidopp 
> so it seems reasonable to add to the controllers a pointer to a DisruptionBudget object, that contains both the Spec (disruption rate limit and max simultaneously down pods) and the Status (recent disruptions).

I was thinking there is no need to add such a pointer, we can just create a `DisruptionBudget`, and use Selector/Label to match Pods. So, in general, the design of DisruptionBudget looks like:

```
// DisruptionBudget is an object to define the max disruption that can be caused to a set of pods
type DisruptionBudget struct {
	unversioned.TypeMeta `json:",inline"`
	// Standard object's metadata.
	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
	ObjectMeta `json:"metadata,omitempty"`
	// Specification of the desired behavior of the DisruptionBudget.
	Spec DisruptionBudgetSpec `json:"spec,omitempty"`
	// Most recently observed status of the DisruptionBudget.
	Status DisruptionBudgetStatus `json:"status,omitempty"`
}

// DisruptionBudgetSpec is a description of a DisruptionBudget.
type DisruptionBudgetSpec struct {
	// Label selector for pods. Existing pods who are selected by this will be the ones affected by this DisruptionBudget.
	Selector *unversioned.LabelSelector `json:"selector,omitempty"`

	// RateLimit is a rate limit on disruptions (preemption and other evictions) across the corresponding
	// set of pods, e.g. no more than one disruption per hour across the pods of a particular Service
	RateLimit RateLimit `json:"ratelimit,omitempty"`

	// The maximum number of pods that can be unavailable simultaneously.
	// Value can be an absolute number (ex: 5) or a percentage of total pods(ex: 10%).
	// Absolute number is calculated from percentage by rounding up.
	// By default, a fixed value of 1 is used.
	MaxUnavailable intstr.IntOrString `json:"maxUnavailable,omitempty"`
}

// DisruptionBudgetStatus represents information about the status of a DisruptionBudget. Status may trail the actual
// state of a system.
type DisruptionBudgetStatus struct {
	// The current disruption rate that the collection of pods is experiencing.
	ObservedBucketTokenNum int `json:"observedBucketTokenNum,omitempty"`
	// Total number of available pods (ready for at least minReadySeconds) targeted by this disruption budget.
	AvailableReplicas int `json:"availableReplicas,omitempty"`
	// Total number of unavailable pods targeted by this disruption budget.
	UnavailableReplicas int `json:"unavailableReplicas,omitempty"`
}

type RateLimit struct {
	BucketCapacity *inf.Dec
	Duration time.Duration
}

// BucketToken is used to limit the rate of pod disruption.
// When disruption(preemption and other evictions) happens, we create a BucketToken(with TTL),
// If the number of BucketTokens in cluster is  not less than BucketCapacity(defined in RateLimit),
// in a preiod of time, we can not disrupt pods in a further step.
type BucketToken struct {
	unversioned.TypeMeta `json:",inline"`
	// Standard object's metadata.
	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
	ObjectMeta `json:"metadata,omitempty"`
	// Specification of this BucketToken.
	Spec BucketTokenSpec
}

type BucketTokenSpec struct{
	// Label selector for pods. Existing pods who are selected by this will be the ones affected by this BucketToken.
	Selector *unversioned.LabelSelector `json:"selector,omitempty"`
}
```

**Comment by mml** [comment]
@mqliang Having budgets match pods by label/selector makes it possible for each pod to be managed by multiple budgets.  We'd have to define the semantics (do we debit all budgets?  do we evict only if all budgets say it's OK?  do we define tie-breaking semantics?).  This is similar to the problem with multiple controllers "pointing" to the same pod, which we also need to resolve with `controllerRef`.  In this instance, I think it would be nice to avoid the problem from the outset.

I like that this also produces fast lookup straight from the pod to its budget, rather than enumerating all budgets and determining which match the given pod.

**Comment by davidopp** [comment]
+1 to @mml's comments above.

**Comment by mqliang** [comment]
@mml Sounds reasonable to me, agree! :+1:

**Comment by mqliang** [comment]
Then, should we put `DisruptionBudget` as a part of Pod, Service or as a part of Controller(RC/RS/Job) putting`DisruptionBudget` as a part of Service sounds reasonable, but how how to handle the case in which people create a Pod directly(without a RC or Service)

**Comment by davidopp** [comment]
DisruptionController is in review; moving to 1.4.

**Comment by goltermann** [comment]
@davidopp can you take a look and bump to v1.5 if not needed for 1.4

**Comment by davidopp** [comment]
This is just waiting on docs.

**Comment by ingvagabund** [comment]
@davidopp what is exactly missing here to document? Devel documents how to implement other DB types (mentioning we handle ``MinAvailable`` for the moment) with explanation what DB is (referring rescheduling docs + additional notes) and how it works based on the current API & controller implementation? Plus with comments from related issues/PRs about current limitations, how is ``PodDisruptionAllowed`` set and how it is (about to be) consumed by ``/eviction`` subresource?

Has anyone started working on that?

**Comment by bgrant0607** [comment]
@mml What remains to be done here?

**Comment by bgrant0607** [comment]
ref #35318, #38336

**Comment by mml** [comment]
I think this is done.
